subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -473,28 +522,130 @@ int ztls_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
 
 ssize_t ztls_send(int sock, const void *buf, size_t len, int flags)
 {
-	return zsock_send(sock, buf, len, flags);
+	return ztls_sendto(sock, buf, len, flags, NULL, 0);
 }
 
 ssize_t ztls_recv(int sock, void *buf, size_t max_len, int flags)
 {
-	return zsock_recv(sock, buf, max_len, flags);
+	return ztls_recvfrom(sock, buf, max_len, flags, NULL, 0);
 }
 
 ssize_t ztls_sendto(int sock, const void *buf, size_t len, int flags,
 		    const struct sockaddr *dest_addr, socklen_t addrlen)
 {
-	return zsock_sendto(sock, buf, len, flags, dest_addr, addrlen);
+	struct net_context *context = INT_TO_POINTER(sock);
+	int ret;
+	bool is_block = true;
+
+	if (!context->tls) {
+		return zsock_sendto(sock, buf, len, flags, dest_addr, addrlen);
+	}
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(context)) {
+		is_block = false;
+	}
+
+	do {
+		k_mutex_lock(&context->tls->mbedtls_lock, K_FOREVER);
```

Comment:
So, in a sense, K_FOREVER contradicts here a case where is_block == false. Overall, all this stuff adds a lot of complication for unknown benefit.